File: backend/diabetesRagAgent.ts
import { StateGraph, CompiledStateGraph, START, END } from "@langchain/langgraph";
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";
import { CheerioWebBaseLoader } from "@langchain/community/document_loaders/web/cheerio";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { GoogleGenerativeAIEmbeddings } from "@langchain/google-genai";
import { formatDocumentsAsString } from "langchain/util/document";
import { transcribeAudioInput, textToSpeech, recordVoice, playAudio } from './utils/audio';
import { getApiKey } from "./utils/env";
import logger from "./utils/logger";
import { Document } from "langchain/document";

// Define our state interface for the agent
export interface DiabetesQnAState {
  question: string;
  category?: "glucose" | "medication" | "meal" | "wellness" | "general";
  relevantDocs?: string;
  needsMoreInfo: boolean;
  answer?: string;
  followupQuestions?: string[];
  conversationHistory?: Array<{ role: string; content: string }>;
}

// Document sources by category
const DOCUMENT_SOURCES = {
  glucose: [
    "https://www.cdc.gov/diabetes/managing/managing-blood-sugar.html",
    "https://www.diabetes.org/healthy-living/medication-treatments/blood-glucose-testing-and-control",
    "https://www.niddk.nih.gov/health-information/diabetes/overview/managing-diabetes/know-blood-sugar-numbers"
  ],
  medication: [
    "https://www.diabetes.org/healthy-living/medication-treatments",
    "https://www.niddk.nih.gov/health-information/diabetes/overview/insulin-medicines-treatments",
    "https://www.cdc.gov/diabetes/managing/medication.html"
  ],
  meal: [
    "https://www.diabetes.org/healthy-living/recipes-nutrition",
    "https://www.cdc.gov/diabetes/managing/eat-well.html",
    "https://www.niddk.nih.gov/health-information/diabetes/overview/diet-eating-physical-activity"
  ],
  wellness: [
    "https://www.diabetes.org/healthy-living/mental-health",
    "https://www.cdc.gov/diabetes/managing/mental-health.html",
    "https://www.niddk.nih.gov/health-information/diabetes/overview/preventing-problems"
  ],
  general: [
    "https://www.diabetes.org/diabetes",
    "https://www.cdc.gov/diabetes/basics/type2.html",
    "https://www.niddk.nih.gov/health-information/diabetes/overview"
  ]
};

/**
 * Class for the Diabetes RAG Agent that handles patient questions
 */
export class DiabetesRagAgent {
  private vectorStores: Record<string, any> = {};
  private model: ChatGoogleGenerativeAI;
  private embeddings: GoogleGenerativeAIEmbeddings;
  private graph: StateGraph<DiabetesQnAState>;
  private executor: any;
  private isInitialized: boolean = false;
  private audioController: AbortController | null = null;

  constructor() {
    const API_KEY = getApiKey("GEMINI_API_KEY");

    this.model = new ChatGoogleGenerativeAI({
      apiKey: API_KEY,
      model: "gemini-2.0-flash",
      maxOutputTokens: 2048,
    });

    this.embeddings = new GoogleGenerativeAIEmbeddings({
      apiKey: API_KEY,
      modelName: "embedding-001",
    });

    this.graph = new StateGraph<DiabetesQnAState>({
      channels: {
        question: null,
        category: null,
        relevantDocs: null,
        needsMoreInfo: null,
        answer: null,
        followupQuestions: null,
        conversationHistory: null
      },
    });
  }

  public async initialize(): Promise<void> {
    if (this.isInitialized) return;

    try {
      logger.info("Initializing Diabetes RAG Agent");
      await this.setupVectorStores();
      this.setupGraph();
      this.executor = this.graph.compile();
      this.isInitialized = true;
      logger.info("Diabetes RAG Agent successfully initialized");
    } catch (error) {
      logger.error("Failed to initialize Diabetes RAG Agent", { error });
      throw new Error(`Failed to initialize RAG system: ${error}`);
    }
  }

  private async setupVectorStores(): Promise<void> {
    const categories = ["glucose", "medication", "meal", "wellness", "general"] as const;

    for (const category of categories) {
      try {
        const webLoaders = DOCUMENT_SOURCES[category].map(
          (source) =>
            new CheerioWebBaseLoader(source, {
              selector: "main, article, .content, .article-content, #content, body",
            })
        );

        const docs: Document[] = [];
        for (const loader of webLoaders) {
          try {
            const loadedDocs = await loader.load();
            docs.push(...loadedDocs);
          } catch (err) {
            logger.warn(`Failed to load a document source, skipping: ${err}`);
          }
        }

        const textSplitter = new RecursiveCharacterTextSplitter({
          chunkSize: 1000,
          chunkOverlap: 200,
        });

        const splitDocs = await textSplitter.splitDocuments(docs);

        this.vectorStores[category] = await MemoryVectorStore.fromDocuments(splitDocs, this.embeddings);

        logger.info(`Processed ${splitDocs.length} document chunks for category: ${category}`);
      } catch (error) {
        logger.error(`Error processing document category: ${category}`, { error });
        this.vectorStores[category] = new MemoryVectorStore(this.embeddings);
      }
    }
  }

  private setupGraph(): void {
    this.graph.addNode("categorize_question", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      const response = await this.model.invoke([
        new SystemMessage(
          "You are an expert at categorizing diabetes-related questions. " +
          "Categorize the given question into one of these categories: " +
          "glucose (blood sugar management), medication (medications and treatments), " +
          "meal (nutrition and diet), wellness (emotional and mental health), " +
          "or general (general diabetes information). " +
          "Respond with only the category name in lowercase."
        ),
        new HumanMessage(state.question),
      ]);

      const category = response.content.toString().trim().toLowerCase();
      const validCategories = ["glucose", "medication", "meal", "wellness", "general"];

      return {
        ...state,
        category: validCategories.includes(category) ? (category as any) : "general",
      };
    });
    
    // Add retrieve_documents node
    this.graph.addNode("retrieve_documents", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      // Default to general if no category is set
      const category = state.category || "general";
      
      // Get the vector store for the category
      const vectorStore = this.vectorStores[category];
      
      try {
        // Perform similarity search
        const docs = await vectorStore.similaritySearch(state.question, 3);
        const relevantDocs = formatDocumentsAsString(docs);
        
        return {
          ...state,
          relevantDocs,
        };
      } catch (error) {
        logger.error(`Error retrieving documents: ${error}`);
        return {
          ...state,
          relevantDocs: "",
          needsMoreInfo: true,
        };
      }
    });
    
    // Add check_info_sufficiency node
    this.graph.addNode("check_info_sufficiency", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      const hasRelevantDocs = state.relevantDocs && state.relevantDocs.trim().length > 0;
      
      return {
        ...state,
        needsMoreInfo: !hasRelevantDocs,
      };
    });
    
    // Add request_more_info node
    this.graph.addNode("request_more_info", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      return {
        ...state,
        answer: "I need more specific information to answer your question about diabetes. Could you provide more details?",
        followupQuestions: [
          "Can you be more specific about what you'd like to know?",
          "Are you asking about a particular aspect of diabetes management?",
          "Would you like information about blood sugar levels, medication, diet, or something else?"
        ]
      };
    });
    
    // Add generate_answer node
    this.graph.addNode("generate_answer", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      const response = await this.model.invoke([
        new SystemMessage(
          "You are a helpful and accurate medical AI assistant for diabetes patients. " +
          "Use the provided context information to answer the question. " +
          "If you don't know the answer, say so rather than making something up. " +
          "Always mention that the patient should consult healthcare professionals for medical advice."
        ),
        new HumanMessage(
          `Context information: ${state.relevantDocs || "No specific information available."}\n\n` +
          `Question: ${state.question}\n\n` +
          `Answer the question based on the context provided.`
        )
      ]);
      
      return {
        ...state,
        answer: response.content.toString(),
      };
    });
    
    // Add generate_followups node
    this.graph.addNode("generate_followups", async (state: DiabetesQnAState): Promise<Partial<DiabetesQnAState>> => {
      if (!state.answer) return { ...state, followupQuestions: [] };
      
      const response = await this.model.invoke([
        new SystemMessage(
          "Based on the user's question and your answer, suggest 3 natural follow-up questions they might want to ask. " +
          "These should be directly related to diabetes management and relevant to the previous conversation."
        ),
        new HumanMessage(
          `User question: ${state.question}\n` +
          `Your answer: ${state.answer}\n` +
          `Generate 3 potential follow-up questions:`
        )
      ]);
      
      // Parse the response to extract the follow-up questions
      const content = response.content.toString();
      const questions = content
        .split(/\d+[\.\)]\s+/)  // Split by numbered list (1. or 1) format)
        .slice(1)  // Remove the first empty item
        .map(q => q.trim())
        .filter(q => q.length > 0);
      
      return {
        ...state,
        followupQuestions: questions.length > 0 ? questions : [
          "What are the common symptoms of diabetes?",
          "How can I monitor my blood sugar at home?",
          "What lifestyle changes can help manage diabetes?"
        ],
      };
    });
    
    // Connect the nodes in the graph
    this.graph.addEdge(START, "categorize_question" as any);
    this.graph.addEdge("categorize_question" as any, "retrieve_documents" as any);
    this.graph.addEdge("retrieve_documents" as any, "check_info_sufficiency" as any);
    this.graph.addConditionalEdges(
      "check_info_sufficiency" as any,
      (state) => (state.needsMoreInfo ? "request_more_info" : "generate_answer")
    );
    this.graph.addEdge("generate_answer" as any, "generate_followups" as any);
    this.graph.addEdge("generate_followups" as any, END);
    this.graph.addEdge("request_more_info" as any, END);
  }

  public async answerQuestion(
    question: string,
    category?: "glucose" | "medication" | "meal" | "wellness",
    conversationHistory?: Array<{ role: string; content: string }>
  ): Promise<{ answer: string; followupQuestions: string[] }> {
    if (!this.isInitialized) {
      await this.initialize();
    }

    const initialState: DiabetesQnAState = {
      question,
      category,
      needsMoreInfo: false,
      conversationHistory,
    };

    const finalState = await this.executor.invoke(initialState);

    return {
      answer: finalState.answer || "I'm sorry, I couldn't generate an answer at this time.",
      followupQuestions: finalState.followupQuestions || [],
    };
  }

  public async handleVoiceInteraction(options: {
    recordingTimeout?: number;
    playbackVolume?: number;
    maxRecordingDuration?: number;
  } = {}): Promise<void> {
    try {
      // Create a new abort controller for this interaction
      this.audioController = new AbortController();
      const signal = this.audioController.signal;

      // Record user's voice input with timeout and cancellation
      const audioBuffer = await recordVoice(options.maxRecordingDuration ?? 5, {
        timeout: options.recordingTimeout ?? 30,
        cancelSignal: signal
      });
      
      // Transcribe the audio to text with timeout and cancellation
      const question = await transcribeAudioInput(audioBuffer, {
        timeout: 30,
        cancelSignal: signal
      });
      
      // Get the answer from the agent
      const { answer } = await this.answerQuestion(question);
      
      // Convert the answer to speech with timeout and cancellation
      const answerAudio = await textToSpeech(answer, {
        timeout: 60,
        cancelSignal: signal
      });
      
      // Play the answer with volume control and cancellation
      await playAudio(answerAudio, {
        volume: options.playbackVolume ?? 1.0,
        cancelSignal: signal
      });
    } catch (error) {
      if (error instanceof AudioError) {
        switch (error.code) {
          case 'CANCELLED':
            logger.info('Voice interaction was cancelled');
            break;
          case 'TIMEOUT':
            logger.warn('Voice interaction timed out');
            break;
          default:
            logger.error('Error in voice interaction:', error);
        }
      } else {
        logger.error('Unexpected error in voice interaction:', error);
      }
      throw error;
    } finally {
      this.audioController = null;
    }
  }

  public cancelVoiceInteraction(): void {
    if (this.audioController) {
      this.audioController.abort();
      this.audioController = null;
    }
  }
}

export const diabetesRagAgent = new DiabetesRagAgent();

File: backend_py/diabetes_rag_agent.py
import os
import csv
import re
import logging
import requests
import numpy as np
from typing import Optional, List, Dict, Any
from dotenv import load_dotenv
from bs4 import BeautifulSoup
import google.generativeai as genai

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

load_dotenv()

DOCUMENT_SOURCES = {
    "glucose": [
        "https://www.diabetes.org/healthy-living/medication-treatments/blood-glucose-testing-and-control",
        "https://www.niddk.nih.gov/health-information/diabetes/overview/managing-diabetes/know-blood-sugar-numbers"
    ],
    "medication": [
        "https://www.diabetes.org/healthy-living/medication-treatments",
        "https://www.niddk.nih.gov/health-information/diabetes/overview/insulin-medicines-treatments"
    ],
    "meal": [
        "https://diabetesjournals.org/care/article/40/Supplement_1/S33/36913/4-Lifestyle-Management",
        "https://www.niddk.nih.gov/health-information/diabetes/overview/diet-eating-physical-activity",
        "./backend_py/data/nutritiondata.csv"
    ],
    "wellness": [
        "https://www.diabetes.org/healthy-living/mental-health",
        "https://www.niddk.nih.gov/health-information/diabetes/overview/preventing-problems"
    ],
    "general": [
        "https://www.cdc.gov/diabetes/about/about-type-2-diabetes.html?CDC_AAref_Val=https://www.cdc.gov/diabetes/basics/type2.html",
        "https://www.niddk.nih.gov/health-information/diabetes/overview",
    ]
}

_VALID_CATEGORIES = ["glucose", "medication", "meal", "wellness", "general"]


class SimpleVectorStore:
    """Lightweight in-memory vector store using cosine similarity."""

    def __init__(self):
        self._embeddings: List[np.ndarray] = []
        self._documents: List[str] = []

    def add_documents(self, documents: List[str], embeddings: List[np.ndarray]) -> None:
        self._documents.extend(documents)
        self._embeddings.extend(embeddings)

    def similarity_search(self, query_embedding: np.ndarray, k: int = 3) -> List[str]:
        if not self._embeddings:
            return []
        matrix = np.vstack(self._embeddings)
        # Normalise query
        q_norm = query_embedding / (np.linalg.norm(query_embedding) + 1e-10)
        # Normalise document embeddings row-wise
        norms = np.linalg.norm(matrix, axis=1, keepdims=True) + 1e-10
        scores = (matrix / norms) @ q_norm
        top_k = min(k, len(self._documents))
        indices = np.argsort(scores)[-top_k:][::-1]
        return [self._documents[i] for i in indices]


def _load_web_document(url: str) -> str:
    """Fetch a URL and return its plain-text content."""
    try:
        headers = {"User-Agent": "Mozilla/5.0 (compatible; DiabetesBot/1.0)"}
        response = requests.get(url, timeout=15, headers=headers)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")
        for tag in soup(["script", "style", "nav", "footer", "header"]):
            tag.decompose()
        text = soup.get_text(separator="\n", strip=True)
        return re.sub(r"\n{3,}", "\n\n", text)
    except Exception as exc:
        logger.warning("Failed to load %s: %s", url, exc)
        return ""


def _load_csv_document(path: str) -> str:
    """Read a CSV file and return its rows as plain text."""
    try:
        rows = []
        with open(path, newline="", encoding="utf-8") as fh:
            reader = csv.DictReader(fh)
            for row in reader:
                rows.append(", ".join(f"{k}: {v}" for k, v in row.items() if v))
        return "\n".join(rows)
    except Exception as exc:
        logger.warning("Failed to load CSV %s: %s", path, exc)
        return ""


def _split_text(text: str, chunk_size: int = 1000, chunk_overlap: int = 200) -> List[str]:
    """Split *text* into overlapping chunks of at most *chunk_size* characters."""
    chunks = []
    start = 0
    while start < len(text):
        chunks.append(text[start:start + chunk_size])
        start += chunk_size - chunk_overlap
    return [c for c in chunks if c.strip()]


class DiabetesRagAgent:
    """Pure-Python RAG agent for diabetes Q&A using Google Gemini directly."""

    def __init__(self):
        api_key = os.getenv("GEMINI_API_KEY")
        genai.configure(api_key=api_key)
        self._model = genai.GenerativeModel("gemini-2.0-flash")
        self._vector_stores: Dict[str, SimpleVectorStore] = {}
        self.is_initialized = False

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def preload_documents(self) -> None:
        self._setup_vector_stores()
        self.is_initialized = True

    def answer_question(
        self,
        question: str,
        category: Optional[str] = None,
        conversation_history: Optional[List[Dict[str, str]]] = None,
    ) -> Dict[str, Any]:
        if not self.is_initialized:
            self.preload_documents()
        cat = self._categorize_question(question, category)
        context = self._retrieve_documents(question, cat)
        answer = self._generate_answer(question, context)
        followups = self._generate_followups(question, answer)
        return {
            "answer": answer or "I'm sorry, I couldn't generate an answer at this time.",
            "followupQuestions": followups,
        }

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _embed_text(self, text: str, task_type: str = "retrieval_document") -> np.ndarray:
        result = genai.embed_content(
            model="models/embedding-001",
            content=text,
            task_type=task_type,
        )
        return np.array(result["embedding"])

    def _setup_vector_stores(self) -> None:
        for category in _VALID_CATEGORIES:
            store = SimpleVectorStore()
            try:
                raw_texts: List[str] = []
                for source in DOCUMENT_SOURCES[category]:
                    if source.endswith(".csv"):
                        logger.info("Loading CSV: %s", source)
                        text = _load_csv_document(source)
                    else:
                        text = _load_web_document(source)
                    if text:
                        raw_texts.append(text)
                        logger.info("✅ Loaded %s (category: %s)", source, category)
                    else:
                        logger.warning("No content from %s", source)

                if not raw_texts:
                    logger.warning("No documents for category: %s", category)
                    self._vector_stores[category] = store
                    continue

                chunks = []
                for text in raw_texts:
                    chunks.extend(_split_text(text))

                if not chunks:
                    logger.warning("No chunks for category: %s", category)
                    self._vector_stores[category] = store
                    continue

                embeddings = [self._embed_text(c) for c in chunks]
                store.add_documents(chunks, embeddings)
                logger.info("Stored %d chunks for category: %s", len(chunks), category)
            except Exception as exc:
                logger.error("Error processing category %s: %s", category, exc)
            self._vector_stores[category] = store

    def _categorize_question(self, question: str, category: Optional[str]) -> str:
        if category and category in _VALID_CATEGORIES:
            logger.info("Skipping categorization — category already set: %s", category)
            return category
        prompt = (
            "You are an expert at categorizing diabetes-related questions. "
            "Categorize the given question into one of these categories: "
            "glucose (blood sugar management), medication (medications and treatments), "
            "meal (nutrition and diet), wellness (emotional and mental health), "
            "or general (general diabetes information). "
            "Respond with only the category name in lowercase.\n\n"
            f"Question: {question}"
        )
        response = self._model.generate_content(prompt)
        cat = response.text.strip().lower()
        return cat if cat in _VALID_CATEGORIES else "general"

    def _retrieve_documents(self, question: str, category: str) -> str:
        store = self._vector_stores.get(category, SimpleVectorStore())
        try:
            query_embedding = self._embed_text(question, task_type="retrieval_query")
            docs = store.similarity_search(query_embedding, k=3)
            logger.info("Retrieved %d documents for category '%s'", len(docs), category)
            return "\n\n".join(docs)
        except Exception as exc:
            logger.error("Error retrieving documents: %s", exc)
            return ""

    def _generate_answer(self, question: str, context: str) -> str:
        prompt = (
            "You are a helpful and accurate medical AI assistant for diabetes patients. "
            "Use the provided context information to answer the question if it is relevant. "
            "If the context does not contain the answer, use your own knowledge to provide the most accurate and helpful response. "
            "Do not say 'I am sorry, but this document does not contain information about ...' or similar phrases. "
            "Always provide a helpful, informative answer, and mention that the patient should consult healthcare professionals for medical advice.\n\n"
            f"Context information: {context or 'No specific information available.'}\n\n"
            f"Question: {question}\n\n"
            "Answer the question based on the context provided, or your own knowledge if the context is insufficient."
        )
        response = self._model.generate_content(prompt)
        return response.text

    def _generate_followups(self, question: str, answer: str) -> List[str]:
        if not answer:
            return []
        prompt = (
            "Based on the user's question and your answer, suggest 1 natural follow-up question the user might want to ask. "
            "It should be directly related to diabetes management and relevant to the previous conversation. "
            "It must be a short question, no more than 10 words.\n\n"
            f"User question: {question}\n"
            f"Your answer: {answer}\n"
            "Generate 1 potential follow-up question:"
        )
        response = self._model.generate_content(prompt)
        content = response.text.replace("**", "")
        questions = [q.strip() for q in content.split("\n") if q.strip()]
        return questions if questions else []


rag_agent = DiabetesRagAgent()
